File: skydo/Utilities/ApiService.swift
//
//  ApiService.swift
//  skydo
//
//  Created by Sebastian Jörz on 28.12.22.
//

import Appwrite
import Foundation

// MARK: - Constants

let dbId = "63ada9616e452192e3e5"
let collId = "63ada9879e426f3b85f5"
let universalLinkVerify = "https://skydo.skyface.de/verify"

let appwriteClient = Client()
  .setEndpoint("https://appwrite.skyface.de/v1")
  .setProject("63ad9f6f86df3acb7d0a")

let database = Databases(appwriteClient)
let account = Account(appwriteClient)

// MARK: - API

struct APIService {
  func getTodos(listArchived: Bool) async throws -> [TodoItem] {
    do {

      let documentList = try await database.listDocuments(
        databaseId: dbId,
        collectionId: collId,
        queries: [
          Query.equal("isActive", value: !listArchived)
        ]
      )

      var todos: [TodoItem] = []

      for document in documentList.documents {
        let decodedDocument = document.convertTo(fromJson: converter)
        todos.append(decodedDocument)
      }

      // Sort by $createdAt: Newest -> Latest
      todos = todos.sorted { a, b in
        return a.createdAt! > b.createdAt!
      }

      // Sort by Completed: Not completed -> Completed
      todos = todos.sorted(by: { a, b in
        return !a.completed
      })

      // Sort by remindTime: Latest -> Newest
      // todos = todos.sorted { a, b in
      //   if a.remindTime == nil || b.remindTime == nil {
      //     return !a.completed
      //   }
      //   return a.remindTime! > b.remindTime!
      // }

      // // Sort by Completed: Not completed -> Completed
      // todos = todos.sorted(by: { a, b in
      //   return !a.completed
      // })
      return todos
    } catch {
      throw error
    }
  }

  static func createTodo(todo: TodoItem, withRemTime: Bool) async throws -> Bool {

    let paramsMap: [String: Any?] = todo.toParamsMap(withRemindTime: withRemTime)

    /* var parameters: [String: Any] = ["title": title, "desc": desc, "completed": false]
    if remTime != nil {
      let formatter = DateFormatter()
      formatter.locale = Locale(identifier: "en_US_POSIX")
      formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
      let dateString = formatter.string(from: remTime!)
      parameters["remindTime"] = dateString
    }*/

    do {
      let result = try await database.createDocument(
        databaseId: dbId, collectionId: collId, documentId: ID.unique(), data: paramsMap)

      print(result)
      return true
    } catch {
      print("ERROR \(error)")
      throw error
    }
  }

  static func toggleActiveTodo(id: String, isActive: Bool) async {

    let parameters: [String: Any] = ["isActive": isActive]
    do {
      let result = try await database.updateDocument(
        databaseId: dbId, collectionId: collId, documentId: id, data: parameters)
      print(result)
    } catch {
      print("ERROR \(error)")
    }
  }

  static func toggleTodoCompletion(id: String, completed: Bool) async {

    let parameters: [String: Any] = ["completed": completed]
    do {
      let result = try await database.updateDocument(
        databaseId: dbId, collectionId: collId, documentId: id, data: parameters)
      print(result)
    } catch {
      print("ERROR \(error)")
    }
  }

  static func logout() async throws {

    do {
      _ = try await account.deleteSessions()
    } catch {
      throw error
    }
  }

  static func updateTodo(
    todo: TodoItem, withRemTime: Bool
  ) async throws {
    let paramsMap: [String: Any?] = todo.toParamsMap(withRemindTime: withRemTime)

    print("Update PARAMS")
    print(paramsMap)
    do {
      _ = try await database.updateDocument(
        databaseId: dbId, collectionId: collId, documentId: todo.id, data: paramsMap)

    } catch {
      throw error
    }
  }

  static func checkLoginStatus() async -> Bool {
    do {
      _ = try await account.get()
      return true
    } catch {
      return false
    }
  }

  static func createVerification() async throws {
    _ = try await account.createVerification(url: universalLinkVerify)
  }

  static func verifyConfirm(userId: String, secret: String) async throws {
    let account = Account(appwriteClient)
    _ = try await account.updateVerification(userId: userId, secret: secret)
  }
}

// MARK: - Helpers
func dateToString(date: Date) -> String {
  let formatter = DateFormatter()
  formatter.locale = Locale(identifier: "en_US_POSIX")
  formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZ"
  let dateString = formatter.string(from: date)
  return dateString
}

File: skydo_flutter/lib/api_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:appwrite/models.dart' as models;
import 'package:skydo_flutter/todos/todo_model.dart';

String dbId = "63ada9616e452192e3e5";
String collId = "63ada9879e426f3b85f5";

class ApiService {
  static Client client = Client();
  static void init() {
    client
        .setEndpoint('https://appwrite.skyface.de/v1')
        .setProject('63ad9f6f86df3acb7d0a')
        .setSelfSigned(status: true);
  }

  static Future<bool> login(String email, String password) async {
    Account account = Account(client);
    try {
      models.Session response =
          await account.createEmailSession(email: email, password: password);
      print(response);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future<List<TodoItem>?> getTodos({bool? withArchive}) async {
    init();
    Databases database = Databases(client);
    try {
      List<String> queries = [];
      if (withArchive != null && !withArchive) {
        queries.add(Query.equal('isActive', true));
      }
      models.DocumentList docs = await database.listDocuments(
          databaseId: dbId, collectionId: collId, queries: queries);

      List<TodoItem> todos = [];
      for (var item in docs.documents) {
        todos.add(TodoItem(
          id: item.$id,
          title: item.data['title'],
          desc: item.data['desc'],
          completed: item.data['completed'],
          isActive: item.data['isActive'],
          remindTime: item.data['remindTime'] != null
              ? DateTime.parse(item.data['remindTime'])
              : null,
          priority: item.data['priority'] != null
              ? todoPriorityFromString(item.data['priority'])
              : TodoPriority.without,
        ));
      }
      // Sort todos by remindTime
      todos.sort((a, b) {
        if (a.remindTime == null && b.remindTime == null) {
          return 0;
        } else if (a.remindTime == null) {
          return 1;
        } else if (b.remindTime == null) {
          return -1;
        } else {
          return a.remindTime!.compareTo(b.remindTime!);
        }
      });
      // Sort todos by completed
      todos.sort((a, b) {
        if (a.completed == b.completed) {
          return 0;
        } else if (a.completed) {
          return 1;
        } else {
          return -1;
        }
      });
      return todos;
    } catch (e) {
      print(e);
      return null;
    }
  }

  static Future<bool> createTodo({
    required String title,
    required String desc,
    DateTime? remindTime,
  }) async {
    init();
    if (title.isEmpty) {
      return false;
    }
    Databases database = Databases(client);
    try {
      models.Document doc = await database.createDocument(
          databaseId: dbId,
          collectionId: collId,
          documentId: ID.unique(),
          data: {
            'title': title,
            'desc': desc,
            'completed': false,
            'isActive': true,
            'remindTime': remindTime?.toIso8601String()
          });
      print(doc);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future<bool> updateTodo({
    required String id,
    required String title,
    required String desc,
    required bool completed,
    DateTime? remindTime,
    required TodoPriority pritority,
  }) async {
    if (title.isEmpty) {
      return false;
    }
    init();
    Databases database = Databases(client);
    try {
      models.Document doc = await database.updateDocument(
          databaseId: dbId,
          collectionId: collId,
          documentId: id,
          data: {
            'title': title,
            'desc': desc,
            'completed': completed,
            'remindTime': remindTime?.toIso8601String(),
            'priority': todoPriorityToString(pritority)
          });
      print(doc);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future<bool> toggleCompleted(String id, bool completed) async {
    init();
    Databases database = Databases(client);
    try {
      models.Document doc = await database.updateDocument(
          databaseId: dbId,
          collectionId: collId,
          documentId: id,
          data: {'completed': completed});
      print(doc);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future<bool> toggleActive(String id, bool isActive) async {
    init();
    Databases database = Databases(client);
    try {
      models.Document doc = await database.updateDocument(
          databaseId: dbId,
          collectionId: collId,
          documentId: id,
          data: {'isActive': isActive});
      print(doc);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future<bool> register(String email, String password) async {
    init();
    Account account = Account(client);
    try {
      models.Account acc = await account.create(
          email: email,
          password: password,
          name: email.split('@')[0],
          userId: ID.unique());
      print(acc);
      return true;
    } catch (e) {
      print(e);
      return false;
    }
  }

  static Future checkLogin() async {
    init();
    Account account = Account(client);
    try {
      models.Account acc = await account.get();
    } catch (e) {
      rethrow;
    }
  }
}
